### PR TITLE
Matrix Cleanup Step 4

### PR DESCRIPTION
**Pull request by siefkenj**
This continues the matrix cleanup work of #12360 and #12042 and #11954 by splitting all determinant-related methods into the `MatrixDeterminant` class.

Both the Bareiss and Berkowitz algorithms have been rewritten in a recursive style (rather than iterative) in `_eval_det_berkowitz` and `_eval_det_bareiss`.

Snake-case names `cofactor_matrix`, `minor`, and `minor_submatrix` have been introduced to supplement `cofactorMatrix`, `minorEntry`, and `minorMatrix`.

`det` now uses a hard-coded formula for matrices smaller than 3x3.

Essential the only changes in this PR are the creation of `MatrixDeterminant` and addition of tests in `test_commonmatrix.py`.  Code coverage for methods of `MatrixDeterminant` is near 100%.

### CODE REVIEW

**Comment by siefkenj** [comment]
@moorepants would you take a look at this?

File: sympy/matrices/matrices.py [hunk 1]
```diff
@@ -1676,7 +1676,371 @@ def multiply_elementwise(self, other):
...
+        elimination method. This approach is best suited for dense
+        symbolic matrices and will result in a determinant with
+        minimal number of fractions. It means that less term
+        rewriting is needed on resulting formulae.
+
+        TODO: Implement algorithm for sparse matrices (SFF),
+        http://www.eecis.udel.edu/~saunders/papers/sffge/it5.ps.
+        """
+
+        # XXX included as a workaround for issue #12362.  Should use `_find_reasonalbe_pivot` instead
```
Comment (smichr) [inline sympy/matrices/matrices.py]: twiddle `lb` in reasonable.

**Comment by smichr** [comment]
FWIW, it's looking good to me.

File: sympy/matrices/matrices.py [hunk 2]
```diff
@@ -1676,7 +1676,371 @@ def multiply_elementwise(self, other):
...
+        if self.rows == 0 and self.cols == 0:
+            return self._new(1, 1, [S.One])
+        elif self.rows == 1 and self.cols == 1:
+            return self._new(2, 1, [S.One, -self[0,0]])
+
+        submat, toeplitz = self._eval_berkowitz_toeplitz_matrix()
+        return toeplitz * submat._eval_berkowitz_vector()
+
+    def _eval_det_bareiss(self):
+        """Compute matrix determinant using Bareis' fraction-free
```
Comment (jksuom) [inline sympy/matrices/matrices.py]: Bareiss'

File: sympy/matrices/matrices.py [hunk 3]
```diff
@@ -1676,7 +1676,371 @@ def multiply_elementwise(self, other):
...
+
+        # XXX included as a workaround for issue #12362.  Should use `_find_reasonable_pivot` instead
+        def _find_pivot(l):
+            for pos,val in enumerate(l):
+                if val:
+                    return (pos, val, None, None)
+            return (None, None, None, None)
+
+
+        # Recursively implimented Bareiss algorithm as per Deanna Richelle Leggett's
```
Comment (jksuom) [inline sympy/matrices/matrices.py]: implemented Bareiss'

File: sympy/matrices/matrices.py [hunk 4]
```diff
@@ -1676,7 +1676,371 @@ def multiply_elementwise(self, other):
...
+        except ValueError:
+            # if there is a value error in the LU decomposition, it means the matrix
+            # wasn't invertible
+            return S.Zero
+
+        # get the sign of the determinant
+        det = (-1) ** (len(p) % 2)
+        # multiply by the diagonal entries
+        for k in range(self.rows):
+            det *= l[k,k] * u[k, k]
```
Comment (jksuom) [inline sympy/matrices/matrices.py]: space in k,k

**Comment by jksuom** [comment]
Looks good to me too. (Only some typos.)

**Comment by siefkenj** [comment]
Great.  Then I think we're good to merge :-).

**Comment by moorepants** [comment]
This PR add a 2X speed reduction in determinant calcs. See:

http://www.moorepants.info/misc/sympy-asv/#solve.TimeMatrixOperations.time_det?commits=a8f20979&idx=5

**Comment by moorepants** [comment]
There is a ~2x speed regression from https://github.com/sympy/sympy/commit/a8f20979 in http://www.moorepants.info/misc/sympy-asv/#solve.TimeMatrixOperations.time_det_bareiss

**Comment by siefkenj** [comment]
This is an interesting case. If I recall correctly, I tested the new implementation on matrices of varying sizes. For larger matrices, the current implementation is faster, but for smaller ones, the old one is. Either way, I think the current implementation is more readable...

Should an alternative implementation be added for use with small matrices? I think 5x5 is where hardcoding the formula is slower than the other det methods, but for 4x4, a hardcoded formula is faster than any of the other methods.

**Pull request merged**